File: 0338-counting-bits/0338-counting-bits.c
#include <stdlib.h>

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* countBits(int n, int* returnSize) {
    int* bits = (int*)malloc(sizeof(int) * (n + 1));

    *returnSize = n + 1;

    bits[0] = 0;

    for (int i = 1; i <= n; i++) {
        bits[i] = bits[i >> 1] + (i & 1);
    }

    return bits;
}

File: 0338-counting-bits/0338-counting-bits.py
class Solution:
    def countBits(self, n: int) -> List[int]:
        bits = [0] * (n + 1)

        for i in range(1, n + 1):
            # i >> 1은 i // 2와 같고,
            # i & 1은 i가 홀수인지 확인하는 값이다.
            bits[i] = bits[i >> 1] + (i & 1)

        return bits